src/posts/flox-smart/index.md

```diff
@@ -0,0 +1,215 @@
+
+These two properties are particularly relevant for "climatology" calculations (e.g. `groupby("time.month").mean()`) — a common Xarray workload in the Earth Sciences.
+
+First, `method="blockwise"` which applies the grouped-reduction in a blockwise fashion.
+This is great for `resample(time="Y").mean()` where we group by `"time.year"`, which is a monotonic increasing array.
+With an appropriate (and usually quite cheap) rechunking, the problem is embarrassingly parallel.
+![blockwise](https://flox.readthedocs.io/en/latest/_images/new-blockwise-annotated.svg)
+
+Second, `method="cohorts"` which is a bit more subtle.
+Consider `groupby("time.month")` for the monthly mean dataset i.e. grouping by an exactly periodic array.
+When the chunk size along the core dimension "time" is a divisor of the period; so either 1, 2, 3, 4, or 6 in this case; groups tend to occur in cohorts ("groups of groups").
+For example, with a chunk size of 4, monthly mean input data for the "cohort" Jan/Feb/Mar/Apr are _always_ in the same chunk, and totally separate from any of the other months.
+Here is a schematic illustration where each month is represented by a different shade of red and a single chunk contains 4 months:
+![monthly cohorts](https://flox.readthedocs.io/en/latest/_images/cohorts-month-chunk4.png)
+This means that we can run the tree reduction for each cohort (three cohorts in total: `JFMA | MJJA | SOND`) independently and expose more parallelism.
+Doing so can significantly reduce memory required for the computation.
+
+Finally, if there isn't much separation of groups into cohorts, like when groups are randomly distributed across chunks, then it's hard to do better than the standard `method="map-reduce"`.
+
+## Choosing a strategy is hard, and harder to teach.
+
+These strategies are great, but the downside is that some sophistication is required to apply them.
+Worse, they are hard to explain conceptually! I've tried! ([example 1](https://discourse.pangeo.io/t/optimizing-climatology-calculation-with-xarray-and-dask/2453/20?u=dcherian), [example 2](https://discourse.pangeo.io/t/understanding-optimal-zarr-chunking-scheme-for-a-climatology/2335)).
+
+What we need is to choose the appropriate strategy automatically.
+
+## Demo
+
+Here's a quick demo of computing monthly mean climatologies with the National Water Model.
+
+For this input dataset, chunked so that approximately a month of data is in a single chunk,
+
+<RawHTML filePath='/posts/flox-smart/dataset-repr.html' />
+
+we run
+
+```python
+mean_mapreduce = ds.groupby("time.month").mean(method="map-reduce")
+mean_cohorts = ds.groupby("time.month").mean() # this is auto-detected!
+```
+
+Using the algorithm described below, flox will **automatically** set
+`method="cohorts"` for this dataset unless specified, yielding a 5X decrease in
+memory used and a 2X increase in runtime. Read on to figure out how!
+Read on to figure out how!
```

Comment:
A significant number of readers will think it's cool that this is now automatic, but not really care about the specifics of how you did it. That section below is also very dense. I think if there is any other information that you want readers to take away from this point onwards (e.g. that `Groupers` exist) you should telegraph it, and explicitly give them the option to skip forwards to a different subheading guilt-free.